src/main/scala/org/broadinstitute/dsde/workbench/sam/dataAccess/PostgresDistributedLockDAO.scala

```diff
@@ -0,0 +1,131 @@
+  }
+
+  def withLock(lock: LockDetails): Resource[F, Unit] = {
+      Resource.make {
+        Stream
+          .retry[F, Unit](acquireLock(lock), config.retryInterval, identity, config.maxRetry, retryable)
+          .adaptError { case e => new WorkbenchException(s"Reached max retry: $e") }
+          .compile
+          .drain
+      }(_ => releaseLock(lock))
+  }
+
+  private[dsde] def acquireLock(lock: LockDetails): F[Unit] = {
+    for {
+      lockStatus <- getLockStatus(lock)
+      _ <- lockStatus match {
+        case Available =>
+          setLock(lock).recoverWith(_ => F.raiseError[Unit](FailToObtainLock(lock)))
+        case Locked    => F.raiseError[Unit](FailToObtainLock(lock))
+      }
+    } yield ()
+  }
+
+  private[dsde] def getLockStatus(lock: LockDetails): F[LockStatus] = {
+    val res = for {
+      _ <- OptionT(F.delay(retrieveLock(lock)))
+    } yield Locked
+    res.fold[LockStatus](Available)(identity)
+  }
+
+  private[dsde] def setLock(lock: LockDetails): F[Unit] = {
+    for {
+      currentTime <- F.realTimeInstant
+      expiration = currentTime.plus(lock.expiresIn.toMillis, ChronoUnit.MILLIS)
+      _ <- F.delay(createLock(lock, expiration))
+    } yield ()
+  }
+
+  private[dsde] def releaseLock(lock: LockDetails): F[Unit] = {
+    F.delay(deleteLock(lock))
+  }
+
+  private[dsde] def retrieveLock(lock: LockDetails): Option[DistributedLockRecord] = {
+    writeDbRef.inLocalTransaction(implicit session => {
+      val dl = DistributedLockTable.syntax("dl")
+      val column = DistributedLockTable.column
+      samsql"""select ${dl.resultAll}
+               from ${DistributedLockTable as dl}
+               where ${column.lockName} = ${lock.lockName} and ${column.lockValue} = ${lock.lockValue}"""
+        .map(DistributedLockTable(dl)).single().apply()
+        .flatMap(record => {
+          if (Instant.now().isAfter(record.expiresAt)) {
+            deleteLockSamSql(lock).update().apply()
+            None
+          } else {
+            Some(record)
+          }
+        })
+    })
+  }
+
+  private[dsde] def createLock(lock: LockDetails, expiresAt: Instant): Int = {
+    writeDbRef.inLocalTransaction(implicit session => {
+      val dl = DistributedLockTable.column
+      val insertLockQuery =
+        samsql"""insert into ${DistributedLockTable.table}
+                 (${dl.lockName}, ${dl.lockValue}, ${dl.expiresAt})
+                 values (${lock.lockName}, ${lock.lockValue}, $expiresAt)"""
+      insertLockQuery.update().apply()
+    })
+  }
+
+  private def deleteLockSamSql(lock: LockDetails): SQL[Nothing, NoExtractor] = {
+    val dl = DistributedLockTable.column
+    samsql"""delete from ${DistributedLockTable.table}
+            where ${dl.lockName} = ${lock.lockName} and ${dl.lockValue} = ${lock.lockValue}"""
+  }
+
+  private[dsde] def deleteLock(lock: LockDetails): Unit = {
+    writeDbRef.inLocalTransaction(implicit session => {
+      deleteLockSamSql(lock).update().apply()
+    })
+  }
+}
```

Comment:
it looks like `deleteLock` is now only used by `releaseLock`. Is it still useful to abstract this out like this or could we merge the two? I think these two functions are small enough (and not too complex) that we could merge them, but I also can see the value in small, simple functions. Which approach is better and why?

Reply:
I'm doing my best to separate the weird Cats lock logic from the more straightforward DAO logic. They could totally get schmunched into one, but I think preserving the logic from the data access makes sense. Plus, it makes it easier to de-catsify in the future!